File: MobileApp/lib/services/socket_service.dart
import 'package:flutter/foundation.dart';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:socket_io_client/socket_io_client.dart'
    as IO;

class SocketService {
  static final SocketService _instance =
      SocketService._internal();

  factory SocketService() => _instance;

  late IO.Socket _socket;
  final ValueNotifier<dynamic> notificationNotifier =
      ValueNotifier(null); // thông báo mới

  bool get isConnected => _socket.connected;

  SocketService._internal() {
    _socket = IO.io(
      dotenv.env['SOCKET_API'] ??
          'ws://192.168.100.248:8080',
      IO.OptionBuilder()
          .setTransports(['websocket'])
          .disableAutoConnect()
          .build(),
    );
  }

  void connect(String userId) {
    if (!_socket.connected) {
      _socket.connect();
    }

    _socket.onConnect((_) {
      print('✅ Connected to server');
      _socket.emit('join_room', userId);

      // Gắn listener duy nhất
      _socket.off('notification');
      _socket.on('notification', (data) {
        print('📩 Notification: $data');
        notificationNotifier.value =
            data; // đẩy giá trị mới
      });
    });

    _socket.onDisconnect(
      (_) => print('🔴 Disconnected from server'),
    );
  }

  void disconnect() {
    _socket.disconnect();
  }
}

File: WebApp/Frontend/src/services/socketService.ts
// socketService.ts
import { io, Socket } from 'socket.io-client';

type NotificationCallback = (data: any) => void;

class SocketService {
    private static instance: SocketService;
    private socket: Socket | null = null;
    private notificationCallback?: NotificationCallback;

    private constructor() { }

    public static getInstance(): SocketService {
        if (!SocketService.instance) {
            SocketService.instance = new SocketService();
        }
        return SocketService.instance;
    }

    public connect(userId: string): void {
        if (!this.socket) {
            this.socket = io(import.meta.env.VITE_SOCKET_API || 'ws://localhost:8080', {
                transports: ['websocket'],
                autoConnect: false,
            });
        }

        if (!this.socket.connected) {
            this.socket.connect();
        }

        this.socket.on('connect', () => {
            console.log('✅ Connected to server');
            this.socket?.emit('join_room', userId);
        });

        this.socket.on('notification', (data) => {
            console.log('📩 Received notification:', data);
            if (this.notificationCallback) {
                this.notificationCallback(data);
            }
        });

        this.socket.on('disconnect', () => {
            console.log('🔴 Disconnected from server');
        });
    }

    public onNotification(callback: NotificationCallback): void {
        this.notificationCallback = callback;
    }

    public offNotification(): void {
        this.notificationCallback = undefined;
    }

    public on(event: string, callback: (data: any) => void): void {
        this.socket?.on(event, callback);
    }

    public emit(event: string, data: any): void {
        this.socket?.emit(event, data);
    }

    public off(event: string): void {
        this.socket?.off(event);
    }

    public disconnect(): void {
        this.socket?.disconnect();
        this.socket = null;
    }

    public get isConnected(): boolean {
        return this.socket?.connected ?? false;
    }

    getSocket() {
        return this.socket;
    }
}

export default SocketService.getInstance();
